classlib/src/main/java/org/teavm/classlib/impl/record/ObjectMethodsSubstitutor.java

```diff
@@ -156,7 +156,9 @@ private ValueEmitter hash(ProgramEmitter pe, ValueType type, ValueEmitter a) {
     private ValueEmitter substituteToString(DynamicCallSite callSite, ProgramEmitter pe) {
         ValueEmitter thisVar = callSite.getArguments().get(0);
         String names = callSite.getBootstrapArguments().get(1).getString();
-        ValueEmitter resultVar = pe.construct(StringBuilder.class, pe.constant("["));
+        ValueEmitter resultVar = pe.construct(StringBuilder.class, pe.constant(((ValueType.Object) thisVar.getType()).getSimpleName()));
```

Comment:
To properly get simple name, you should first acquire ClassReaderSource from DynamicCallSite, then use it to find ClassReader by name and get its simple name. If it's null, try to extract it from fully qualified name using current approach.